File: src/fr/geobert/radis/ui/editor/InfoManagerDialog.java
package fr.geobert.radis.ui.editor;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import fr.geobert.radis.R;
import fr.geobert.radis.db.DbContentProvider;
import fr.geobert.radis.db.InfoTables;
import fr.geobert.radis.tools.Tools;

import java.util.HashMap;

public class InfoManagerDialog extends DialogFragment {
    public static final int THIRD_PARTIES_DIALOG_ID = 1;
    public static final int TAGS_DIALOG_ID = 2;
    public static final int MODES_DIALOG_ID = 3;
    public static final int EDIT_THIRD_PARTY_DIALOG_ID = 4;
    public static final int EDIT_TAG_DIALOG_ID = 5;
    public static final int EDIT_MODE_DIALOG_ID = 6;
    public static final int DELETE_THIRD_PARTY_DIALOG_ID = 7;
    public static final int DELETE_TAG_DIALOG_ID = 8;
    public static final int DELETE_MODE_DIALOG_ID = 9;
    private static HashMap<String, InfoManager> mInfoManagersMap = new HashMap<String, InfoManager>();
    private static HashMap<String, InfoManagerDialog> mInfoManagerDialogMap = new HashMap<String, InfoManagerDialog>();
    private InfoManager mInfoManager = null;
    private int mMode = -1;

    public static void resetInfoManager() {
        mInfoManagersMap.clear();
    }

    public static InfoManagerDialog newInstance(Uri table, String colName,
                                                String title, int editId, int deleteId, int mode) {
        InfoManagerDialog frag = new InfoManagerDialog();
        Bundle args = new Bundle();
        args.putParcelable("table", table);
        args.putString("colName", colName);
        args.putString("title", title);
        args.putInt("editId", editId);
        args.putInt("deleteId", deleteId);
        args.putInt("mode", mode);
        frag.setArguments(args);
        return frag;
    }

    public static InfoManagerDialog createThirdPartiesListDialog(Context ctx) {
        final String k = DbContentProvider.THIRD_PARTY_URI.toString() + "_list";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.THIRD_PARTY_URI,
                    InfoTables.KEY_THIRD_PARTY_NAME,
                    ctx.getString(R.string.third_parties),
                    EDIT_THIRD_PARTY_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID, THIRD_PARTIES_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    public static InfoManagerDialog createTagsListDialog(Context ctx) {
        final String k = DbContentProvider.TAGS_URI.toString() + "_list";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.TAGS_URI,
                    InfoTables.KEY_TAG_NAME, ctx.getString(R.string.tags),
                    EDIT_TAG_DIALOG_ID, DELETE_TAG_DIALOG_ID, TAGS_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    public static InfoManagerDialog createModesListDialog(Context ctx) {
        final String k = DbContentProvider.MODES_URI.toString() + "_list";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.MODES_URI,
                    InfoTables.KEY_MODE_NAME, ctx.getString(R.string.modes),
                    EDIT_MODE_DIALOG_ID, DELETE_MODE_DIALOG_ID, MODES_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    protected static InfoManagerDialog createThirdPartiesDeleteDialog(Context ctx) {
        final String k = DbContentProvider.THIRD_PARTY_URI.toString() + "_del";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.THIRD_PARTY_URI,
                    InfoTables.KEY_THIRD_PARTY_NAME,
                    ctx.getString(R.string.third_parties),
                    EDIT_THIRD_PARTY_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    protected static InfoManagerDialog createTagsDeleteDialog(Context ctx) {
        final String k = DbContentProvider.TAGS_URI.toString() + "_del";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.TAGS_URI,
                    InfoTables.KEY_TAG_NAME, ctx.getString(R.string.tags),
                    EDIT_TAG_DIALOG_ID, DELETE_TAG_DIALOG_ID, DELETE_TAG_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    protected static InfoManagerDialog createModesDeleteDialog(Context ctx) {
        final String k = DbContentProvider.MODES_URI.toString() + "_del";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.MODES_URI,
                    InfoTables.KEY_MODE_NAME, ctx.getString(R.string.modes),
                    EDIT_MODE_DIALOG_ID, DELETE_MODE_DIALOG_ID, DELETE_MODE_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    public static InfoManagerDialog createInfoDeleteDialog(final int deleteId, final Context ctx) {
        switch (deleteId) {
            case DELETE_THIRD_PARTY_DIALOG_ID:
                return createThirdPartiesDeleteDialog(ctx);
            case DELETE_TAG_DIALOG_ID:
                return createTagsDeleteDialog(ctx);
            case DELETE_MODE_DIALOG_ID:
                return createModesDeleteDialog(ctx);
            default:
                return null;
        }
    }

    public static InfoManagerDialog createInfoEditDialog(final int editId, final Context ctx) {
        switch (editId) {
            case EDIT_THIRD_PARTY_DIALOG_ID:
                return createThirdPartiesEditDialog(ctx);
            case EDIT_TAG_DIALOG_ID:
                return createTagsEditDialog(ctx);
            case EDIT_MODE_DIALOG_ID:
                return createModesEditDialog(ctx);
            default:
                return null;
        }
    }

    protected static InfoManagerDialog createThirdPartiesEditDialog(Context ctx) {
        final String k = DbContentProvider.THIRD_PARTY_URI.toString() + "_edit";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.THIRD_PARTY_URI,
                    InfoTables.KEY_THIRD_PARTY_NAME,
                    ctx.getString(R.string.third_parties),
                    EDIT_THIRD_PARTY_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID, EDIT_THIRD_PARTY_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    protected static InfoManagerDialog createTagsEditDialog(Context ctx) {
        final String k = DbContentProvider.TAGS_URI.toString() + "_edit";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.TAGS_URI,
                    InfoTables.KEY_TAG_NAME, ctx.getString(R.string.tags),
                    EDIT_TAG_DIALOG_ID, DELETE_TAG_DIALOG_ID, EDIT_TAG_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    protected static InfoManagerDialog createModesEditDialog(Context ctx) {
        final String k = DbContentProvider.MODES_URI.toString() + "_edit";
        InfoManagerDialog d = mInfoManagerDialogMap.get(k);
        if (d == null) {
            d = InfoManagerDialog.newInstance(DbContentProvider.MODES_URI,
                    InfoTables.KEY_MODE_NAME, ctx.getString(R.string.modes),
                    EDIT_MODE_DIALOG_ID, DELETE_MODE_DIALOG_ID, EDIT_MODE_DIALOG_ID);
            mInfoManagerDialogMap.put(k, d);
        }
        return d;
    }

    private InfoManager createInfoManagerIfNeeded(Uri table, String colName,
                                                  String title, int editId, int deleteId) {
        InfoManager i = mInfoManagersMap.get(table.toString());
        if (null == i) {
            i = new InfoManager(this, title, table, colName, editId, deleteId);
            mInfoManagersMap.put(table.toString(), i);
        }
        return i;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle args = getArguments();
        final Uri table = args.getParcelable("table");
        final String colName = args.getString("colName");
        final String title = args.getString("title");
        final int editId = args.getInt("editId");
        final int deleteId = args.getInt("deleteId");
        final int mode = args.getInt("mode");
        mMode = mode;
        mInfoManager = createInfoManagerIfNeeded(table, colName, title, editId, deleteId);
        switch (mode) {
            case THIRD_PARTIES_DIALOG_ID:
            case TAGS_DIALOG_ID:
            case MODES_DIALOG_ID:
                return mInfoManager.getListDialog();
            case EDIT_MODE_DIALOG_ID:
            case EDIT_TAG_DIALOG_ID:
            case EDIT_THIRD_PARTY_DIALOG_ID:
                return mInfoManager.getEditDialog();
            case DELETE_MODE_DIALOG_ID:
            case DELETE_TAG_DIALOG_ID:
            case DELETE_THIRD_PARTY_DIALOG_ID:
                return Tools.createDeleteConfirmationDialog(getActivity(), new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        createInfoManagerIfNeeded(table, colName, title, editId, deleteId).deleteInfo();
                    }
                });
            default:
                // should not happen
                return null;
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        Dialog d = getDialog();
        switch (mMode) {
            case THIRD_PARTIES_DIALOG_ID:
            case TAGS_DIALOG_ID:
            case MODES_DIALOG_ID:
                mInfoManager.onPrepareDialog((AlertDialog) d);
                break;
            case EDIT_MODE_DIALOG_ID:
            case EDIT_TAG_DIALOG_ID:
            case EDIT_THIRD_PARTY_DIALOG_ID:
                mInfoManager.initEditDialog(d);
                break;
            case DELETE_MODE_DIALOG_ID:
            case DELETE_TAG_DIALOG_ID:
            case DELETE_THIRD_PARTY_DIALOG_ID:
                break;
            default:
                // should not happen
        }
    }
}

File: app/src/main/kotlin/fr/geobert/radis/ui/editor/InfoManagerDialog.kt
package fr.geobert.radis.ui.editor

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.DialogFragment
import fr.geobert.radis.R
import fr.geobert.radis.db.DbContentProvider
import fr.geobert.radis.db.InfoTables
import fr.geobert.radis.tools.Tools
import java.util.*

public class InfoManagerDialog : DialogFragment() {
    private var mInfoManager: InfoManager? = null
    private var mMode = -1

    private fun createInfoManagerIfNeeded(table: Uri, colName: String, title: String, editId: Int, deleteId: Int):
            InfoManager {
        var i: InfoManager? = mInfoManagersMap.get(table.toString())
        if (null == i) {
            i = InfoManager(this, title, table, colName, editId, deleteId)
            mInfoManagersMap.put(table.toString(), i)
        }
        return i
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val args = arguments
        val table = args.getParcelable<Uri>("table")
        val colName = args.getString("colName")
        val title = args.getString("title")
        val editId = args.getInt("editId")
        val deleteId = args.getInt("deleteId")
        val mode = args.getInt("mode")
        mMode = mode
        val infoManager = createInfoManagerIfNeeded(table, colName, title, editId, deleteId)
        mInfoManager = infoManager
        when (mode) {
            THIRD_PARTIES_DIALOG_ID, TAGS_DIALOG_ID, MODES_DIALOG_ID -> return infoManager.getListDialog()
            EDIT_MODE_DIALOG_ID, EDIT_TAG_DIALOG_ID, EDIT_THIRD_PARTY_DIALOG_ID -> return infoManager.getEditDialog()
            DELETE_MODE_DIALOG_ID, DELETE_TAG_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID ->
                return Tools.createDeleteConfirmationDialog(activity, { d, i ->
                    createInfoManagerIfNeeded(table, colName, title, editId, deleteId).deleteInfo()
                })
            else -> // should not happen
                throw RuntimeException("Unknown dialog:$mode")
        }
    }

    override fun onResume() {
        super.onResume()
        val d = dialog
        when (mMode) {
            THIRD_PARTIES_DIALOG_ID, TAGS_DIALOG_ID, MODES_DIALOG_ID -> mInfoManager!!.onPrepareDialog(d as AlertDialog)
            EDIT_MODE_DIALOG_ID, EDIT_TAG_DIALOG_ID, EDIT_THIRD_PARTY_DIALOG_ID -> mInfoManager!!.initEditDialog(d)
            DELETE_MODE_DIALOG_ID, DELETE_TAG_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID -> {
            }
        }// should not happen
    }

    companion object {
        public val THIRD_PARTIES_DIALOG_ID: Int = 1
        public val TAGS_DIALOG_ID: Int = 2
        public val MODES_DIALOG_ID: Int = 3
        public val EDIT_THIRD_PARTY_DIALOG_ID: Int = 4
        public val EDIT_TAG_DIALOG_ID: Int = 5
        public val EDIT_MODE_DIALOG_ID: Int = 6
        public val DELETE_THIRD_PARTY_DIALOG_ID: Int = 7
        public val DELETE_TAG_DIALOG_ID: Int = 8
        public val DELETE_MODE_DIALOG_ID: Int = 9
        private val mInfoManagersMap = HashMap<String, InfoManager>()
        private val mInfoManagerDialogMap = HashMap<String, InfoManagerDialog>()

        public fun resetInfoManager() {
            mInfoManagerDialogMap.clear()
            mInfoManagersMap.clear()
        }

        private fun newInstance(table: Uri, colName: String, title: String, editId: Int, deleteId: Int, mode: Int):
                InfoManagerDialog {
            val frag = InfoManagerDialog()
            val args = Bundle()
            args.putParcelable("table", table)
            args.putString("colName", colName)
            args.putString("title", title)
            args.putInt("editId", editId)
            args.putInt("deleteId", deleteId)
            args.putInt("mode", mode)
            frag.arguments = args
            return frag
        }

        public fun createThirdPartiesListDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.THIRD_PARTY_URI.toString() + "_list"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.THIRD_PARTY_URI, InfoTables.KEY_THIRD_PARTY_NAME,
                        ctx.getString(R.string.third_parties), EDIT_THIRD_PARTY_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID,
                        THIRD_PARTIES_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }

        public fun createTagsListDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.TAGS_URI.toString() + "_list"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.TAGS_URI, InfoTables.KEY_TAG_NAME,
                        ctx.getString(R.string.tags), EDIT_TAG_DIALOG_ID, DELETE_TAG_DIALOG_ID, TAGS_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }

        public fun createModesListDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.MODES_URI.toString() + "_list"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.MODES_URI, InfoTables.KEY_MODE_NAME,
                        ctx.getString(R.string.modes), EDIT_MODE_DIALOG_ID, DELETE_MODE_DIALOG_ID, MODES_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }

        protected fun createThirdPartiesDeleteDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.THIRD_PARTY_URI.toString() + "_del"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.THIRD_PARTY_URI, InfoTables.KEY_THIRD_PARTY_NAME,
                        ctx.getString(R.string.third_parties), EDIT_THIRD_PARTY_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID,
                        DELETE_THIRD_PARTY_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }

        protected fun createTagsDeleteDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.TAGS_URI.toString() + "_del"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.TAGS_URI, InfoTables.KEY_TAG_NAME,
                        ctx.getString(R.string.tags), EDIT_TAG_DIALOG_ID, DELETE_TAG_DIALOG_ID, DELETE_TAG_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }

        protected fun createModesDeleteDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.MODES_URI.toString() + "_del"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.MODES_URI, InfoTables.KEY_MODE_NAME,
                        ctx.getString(R.string.modes), EDIT_MODE_DIALOG_ID, DELETE_MODE_DIALOG_ID,
                        DELETE_MODE_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }

        public fun createInfoDeleteDialog(deleteId: Int, ctx: Context): InfoManagerDialog? {
            when (deleteId) {
                DELETE_THIRD_PARTY_DIALOG_ID -> return createThirdPartiesDeleteDialog(ctx)
                DELETE_TAG_DIALOG_ID -> return createTagsDeleteDialog(ctx)
                DELETE_MODE_DIALOG_ID -> return createModesDeleteDialog(ctx)
                else -> return null
            }
        }

        public fun createInfoEditDialog(editId: Int, ctx: Context): InfoManagerDialog? {
            when (editId) {
                EDIT_THIRD_PARTY_DIALOG_ID -> return createThirdPartiesEditDialog(ctx)
                EDIT_TAG_DIALOG_ID -> return createTagsEditDialog(ctx)
                EDIT_MODE_DIALOG_ID -> return createModesEditDialog(ctx)
                else -> return null
            }
        }

        protected fun createThirdPartiesEditDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.THIRD_PARTY_URI.toString() + "_edit"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.THIRD_PARTY_URI, InfoTables.KEY_THIRD_PARTY_NAME,
                        ctx.getString(R.string.third_parties), EDIT_THIRD_PARTY_DIALOG_ID, DELETE_THIRD_PARTY_DIALOG_ID,
                        EDIT_THIRD_PARTY_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }

        protected fun createTagsEditDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.TAGS_URI.toString() + "_edit"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.TAGS_URI, InfoTables.KEY_TAG_NAME,
                        ctx.getString(R.string.tags), EDIT_TAG_DIALOG_ID, DELETE_TAG_DIALOG_ID, EDIT_TAG_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }

        protected fun createModesEditDialog(ctx: Context): InfoManagerDialog {
            val k = DbContentProvider.MODES_URI.toString() + "_edit"
            var d: InfoManagerDialog? = mInfoManagerDialogMap.get(k)
            if (d == null) {
                d = InfoManagerDialog.newInstance(DbContentProvider.MODES_URI, InfoTables.KEY_MODE_NAME,
                        ctx.getString(R.string.modes), EDIT_MODE_DIALOG_ID, DELETE_MODE_DIALOG_ID, EDIT_MODE_DIALOG_ID)
                mInfoManagerDialogMap.put(k, d)
            }
            return d
        }
    }
}
